JustSaying/AwsTools/MessageHandling/SqsNotificationListener.cs

```diff
@@ -210,6 +215,38 @@ internal async Task ListenLoopAsync(CancellationToken ct)
             }
         }
 
+        /// <summary>
+        /// Creates a task that extends the message visibility of the batch on a cadence of half the timeout duration, until all of the messages have been processed.
+        /// </summary>
+        /// <param name="inflightTracker"></param>
+        /// <param name="cancellationToken"></param>
+        private void StartBatchHeartbeat(MessageBatchInflightTracker inflightTracker, CancellationToken cancellationToken)
+        {
+            _ = Task.Run(async () =>
```

Comment:
I might be missing the key context here.
I thought your cancellation PR was to address a problem where you'd be inundated with messages that are already being processed but backed up. A sort of perfect storm under heavy load, this would fix that as well.

If you were already in a bad state with theadpool starvation for example, then this code may not run reliably and you'll be doomed as well. But with this and an appropriate throttling strategy, I think you'd be very safe, thats before looking at backoff strategies, which can also reduce the risk of an influx of messages.

Reply:
So the root issue from #546 that lead to #571 was that the throttle was failing to make headway through a contended lock. Because of the fire-and-forget nature of the dispatch, the message loop was unaware of this and was just happily grabbing more and more messages because it _looked like_ there wasn't an backlog. This just made the problem worse and worse and generated a memory leak as the `Task` queue backed up. The worker was still doing _some_ processing, so it wasn't unhealthy enough to be terminated or anything, it was just slowly accumulating stuff in the queue which would have eventually brought it down from memory exhaustion.

#571 tries to mitigate that scenario by trying to free up resources by abandoning queued tasks when they are started by "short-circuiting" and returning if it looks like the task was scheduled so long ago as to have likely been considered abandoned by AWS and re-delivered anyway.

Reply:
I'm still struggling to understand what situation #571 would solve. If there still exists an issue where JustSaying keeps merrily getting messages when a worker is at max capacity there is a problem with the listen loop / throttling strategy that needs a head-on fix.

If there is a local queue that's being populated, either by the throttling strategy or by the new sequential processing mode, then I could see you could get stale messages just before the point of processing them, but letting the visibility lock expire is dangerous cause eventually they're going to end up in a dead-letter-queue under normal high-load circumstances, so #571 would be masking the issue IMO.

Reply:
The fix for that issue was #546 so that the throttle works properly,

#571 is trying to abandon work that's likely _already_ been abandoned by AWS (it could be made more conservative to not process the messages if a larger period that _exactly_ the visibility timeout is exceeded, sure) so that it tries to shed the excess if a similar situation happened again so it could maybe catch-up and recover when the load decreases.

In the #546 scenario, it was so bad it _never_ caught up. At 8am the next morning the worker still had a Task queue backlog in the hundreds of messages that were received and queued at least 8 hours previously.

Reply:
Let's say #546 issue wasn't fixed, or there was a similar bug that arose, with this PR (and a max timeout yet to be added here) let's say a worker has 40 messages inflight and none are getting worked through, we'd keep renewing the visibility timeout for those 40 so that no other worker gets them, so we avoid the same excess you describe.

If we have at this point accumulated 3000 messages in our SQS queue, which is better?
- Each work grabs a handful of messages, but drops most back into the queue. At some point they end up in the error queue. (PR #571)
- Each worker keeps hold of the messages they are trying to process, the queue remains relatively static with the messages it has built up. (PR #574)

Neither solve the root issue that #546 solved (which was some great work btw), but for regular day to day processing, the later option is more desirable in my opinion. #571 favours the health of a worker over the health of the system, it doesn't solve it, neither does #574.

Reply:
But there are lots more nuances around this, this PR favours ordering of message processing, so you get more head-of-line blocking 🤷‍♂

Reply:
The only thing I don't like about #574 is that it might cause confusion about why certain messages appear to be just sat stuck in the worker perpetually, rather than being moved onto a mechanism where you'd think problems with the messages would be noticed.

In the case of the issue we were having, we had to infer the messaging impact based on the memory leak (as stuff was being processed eventually anyway); locally there was just lots of wasted resource from holding on to messages that had actually already been dealt with.

Based on our observations, the dead-letter stuff wasn't happening, as the messages were being dealt with by the system as a whole as a black box made of multiple instances. We had no backed up SQS queues, dead letters or messages in error queues.

#571 is trying to acknowledge this locally by freeing resources based on, what albeit are assumptions, that the rest of the system if dealing with the messages it's not dealing with itself in aggregate.

Reply:
There has to be a `ProcessingTimeout` or similar concept in this PR with a sensible default, and logging too.

I'm trying to think if we can combine these ideas in some way, perhaps this PR should keep track of the visibility renews and track a per-message expiry, then we could add a defensive check before invoking the handler and say "ok, we weren't able to extend the timeout here, someone else is probably working on this now". It's a few more moving parts, but would that satisfy your concern?

Currently #571 isn't calculating the expiry quite right btw, we'd need to take the time that the worker received the message and add on the queue visibility timeout, unless I'm reading it wrong.

Reply:
Yes, you're right the receipt time is recorded here:

https://github.com/justeat/JustSaying/blob/33b72585987c88d6960936305e6392de11dc7be1/JustSaying/AwsTools/MessageHandling/SqsNotificationListener.cs#L187

But the visibility timeout isn't being added here:

https://github.com/justeat/JustSaying/blob/33b72585987c88d6960936305e6392de11dc7be1/JustSaying/AwsTools/MessageHandling/SqsNotificationListener.cs#L312